### ISSUE TITLE
script[defer] doesn't work in IE<=9 

### ISSUE THREAD

**Issue by paulirish** [h5bp/lazyweb-requests#42]
_(Edited 2012.03.01)_

**TL;DR**: don't use `defer` for external scripts that can depend on eachother if you need IE <= 9 support

There is a bug in IE<=9 (confirmed, below, by an IE engineer) where if you have two scripts such as...

``` html
<script defer src="jquery.js"></script>
<script defer src="jquery-ui.js"></script>
```

And the first script modifies the dom with `appendChild`, `innerHTML`, (etc.),  the second script can start executing before the first one has finished.  Thus, a dependency between the two will break.

The details of this limitation begin [at this comment](https://github.com/h5bp/lazyweb-requests/issues/42#issuecomment-1901803)

This essentially means that `script[defer]` cannot be used in most cases unless you have dropped IE8 and IE9 support. If, however, you can UA sniff to serve script[defer] to all browsers except IE6-9, that will net you large performance wins.

Steve Souders indicated there may be a hack of inserting an empty `<script></script>` tag between the two tags that may address this problem. Research to be done…

_original post follows:_

<hr><hr><hr>
# comprehensive research and article on script @defer

`@defer` scripts execute when the browser gets around to them, but they execute in order. this is awesome for performance.
it's also awesome that it's been in IE since IE5.

but, we're lacking a little bit of comprehensive research on this..

kyle simpson thinks there may be some edge case issues with defer... from [this h5bp thread](https://github.com/paulirish/html5-boilerplate/issues/28)...

> 1. support of defer on dynamic script elements isn't defined or supported in any browser... only works for script tags in the markup. this means it's completely useless for the "on-demand" or "lazy-loading" techniques and use-cases.
> 2. i believe there was a case where in some browsers defer'd scripts would start executing immediately before DOM-ready was to fire, and in others, it happened immediately after DOM-ready fired. Will need to do more digging for more specifics on that.
> 3. `defer` used on a script tag referencing an external resource behaved differently than `defer` specified on a script tag with inline code in it. That is, it couldn't be guaranteed to work to defer both types of scripts and have them still run in the correct order.
> 4. `defer` on a script tag written out by a `document.write()` statement differed from a script tag in markup with `@defer`.

it'd be excellent to get a great summary of the full story across browsers and these issues so we can use defer confidently.

see also:
- http://stevesouders.com/hpws/js-defer.php
- http://stevesouders.com/efws/couple-script-defer.php
- http://hacks.mozilla.org/2009/06/defer/

<hr>
- @aaronpeters @Schepp

**Comment by DrDoorknob** [comment]
Because my organization supports IE9 and I've been asked to look into this, I figure I should throw in a comment even if it's a year ahead. I just put together a test of this in an IE9 VM, and couldn't reproduce the issue discussed; in fact I'm currently considering simply universally adding the "defer" attribute to our scripts.
What could potentially be causing the difference in behavior is that anyone working on this issue is trying to find very, very large javascript files that will always take longer to load than the small script below (which requires the large). JQuery is a popular example, and I'm now wondering if that happens to be doing something that causes glitchy behavior to just IE9; changing the DOM being a possible candidate. My test script was a 1.5MB custom pre-built bundle of the dojo framework, and many AMD module definitions; it's likely that script doesn't do anything to the page until DOMContentLoaded, so it seemed to be working okay.
To try to isolate whether frameworks like jquery are actually specifically relevant to the issue, another test could just involve a 2MB javascript file that just contains many random characters, does nothing, and then sets a flag on window.

**Comment by getify** [comment]
> so it seemed to be working okay.

I wouldn't build my production strategy on such false assumptions, but YMMV. As far as I'm concerned, everything here about `defer` is still valid until IE9 stops being supported.

Even in IE10+, `defer` isn't fully reliable, as it will not work on inline scripts, which is unfortunately still a pretty common occurrence, even for just simple code like activating a social widget or GA tracking script, etc.

**Comment by DrDoorknob** [comment]
Didn't you just write this one earlier in the thread?
http://www.whatwg.org/specs/web-apps/current-work/multipage/scripting-1.html#attr-script-defer
"The defer and async attributes must not be specified if the src attribute is not present."

Any syncronicity between a defer+src script with an inline script (ie, something running a JQuery function) is coincidential, and undefined browser behavior.

I tend to think of inline scripts (as in, inline between pieces of content, not a single inline script in head that configures your libraries) as archaic deprecated behavior, like conditional HTML comments; , so I'm not surprised that having them in place breaks newer script-loading features.

**Comment by mamunbd24** [comment]
Hello,

what is the difference between <script async=''async''   & <script async=''??

**Comment by DrDoorknob** [comment]
None, really...they're both syntactically accepted because some HTML-writing programs don't follow the Boolean nature of attributes, and because it's required for valid XML.

Attributes like "async" or "disabled" don't actually require an equals sign. It's valid to write <script async> or <input disabled>. Anything inside the equals is pretty much ignored. When checking for those Booleans, JS code usually checks "element.hasAttribute(x)" rather than "element.getAttribute()", so even if it registers as an empty string, it will turn out true.

**Comment by mamunbd24** [comment]
Thank you for your answer. So wht do you think about the following code?

<script async> src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
<script>
  (adsbygoogle = window.adsbygoogle || []).push({
    google_ad_client: "ca-pub-xxxxxxxxxxxxxxxx",
    enable_page_level_ads: true
  });
</script>

**Comment by mamunbd24** [comment]
I tired it but it shows error message like following: 

Error parsing XML, line 5, column 14: Attribute name "async" associated with an element type "script" must be followed by the ' = ' character.

**Comment by getify** [comment]
For posterity sake, let's not clutter up this thread with general support inquiries about web platform features. Please take this discussion to stackoverflow or some other forum where it's more appropriate, and remove the messages here.

**Comment by vlakoff** [comment]
An user reported the [suggested fix](https://github.com/h5bp/lazyweb-requests/issues/42#issuecomment-26034923) didn't work for him. If other users could confirm this, the original post should be edited so that people who land here are not misled ;)
